tests/rust/safe_abstraction/mutex_u32.rs

```diff
@@ -0,0 +1,444 @@
+    &*& [_]frac_borrow(km, MutexU32_frac_borrow_content(klong, t, lock))
+    &*& SysMutex_locked(&(*lock).inner, full_borrow_wrapper(klong, u32_full_borrow_content(t, &(*lock).data)), t);
+
+pred_ctor MutexGuardU32_full_borrow_content0(km: lifetime_t, t: thread_id_t, l: *MutexGuardU32)() =
+    (*l).lock |-> ?lock &*& MutexGuardU32_own(km)(t, lock);
+
+pred_ctor MutexGuardU32_share(km: lifetime_t)(k: lifetime_t, t: thread_id_t, l: *MutexGuardU32) = true;
+
+lem MutexGuardU32_share_mono(k: lifetime_t, k1: lifetime_t, t: thread_id_t, l: *MutexGuardU32)
+    req lifetime_inclusion(k1, k) == true &*& exists(?km) &*& [_]MutexGuardU32_share(km)(k, t, l);
+    ens [_]MutexGuardU32_share(km)(k1, t, l);
+{
+    close MutexGuardU32_share(km)(k1, t, l);
+    leak MutexGuardU32_share(km)(k1, t, l);
+}
+
+lem MutexGuardU32_full_share(k: lifetime_t, t: thread_id_t, l: *MutexGuardU32)
+    req exists(?km) &*& full_borrow(k, MutexGuardU32_full_borrow_content0(km, t, l)) &*& [?q]lifetime_token(k);
+    ens [_]MutexGuardU32_share(km)(k, t, l) &*& [q]lifetime_token(k);
+{
+    leak full_borrow(_, _);
+    close MutexGuardU32_share(km)(k, t, l);
+    leak MutexGuardU32_share(km)(k, t, l);
+}
+@*/
+impl MutexGuardU32 {
+    unsafe fn new<'a>(lock: &'a MutexU32) -> MutexGuardU32
+/*@ req thread_token(?t) &*& [?qa]lifetime_token(?a) &*& [_]exists_np(?km) &*& lifetime_inclusion(a, km) == true
+    &*& [_]frac_borrow(a, MutexU32_frac_borrow_content(km, t, lock))
+    &*& SysMutex_locked(&(*lock).inner, full_borrow_wrapper(km, u32_full_borrow_content(t, &(*lock).data)), t)
+    &*& full_borrow(km, u32_full_borrow_content(t, &(*lock).data));
+    @*/
+    //@ ens thread_token(t) &*& [qa]lifetime_token(a) &*& MutexGuardU32_own(a)(t, lock);
+    {
+        //@ close MutexGuardU32_own(a)(t, lock);
+        MutexGuardU32 { lock }
+    }
+
+    /*
+    Todo: deref_mut should be in the implementation of the trait `DerefMut`. Support for the implementation for standard library traits is
+    needed for that.
+    Todo: deref_mut should be a safe function */
+    unsafe fn deref_mut<'a>(&'a mut self) -> &'a mut u32
+    /*@ req thread_token(?t) &*& [?qa]lifetime_token(?a) &*& exists(?km)
+        &*& full_borrow(a, MutexGuardU32_full_borrow_content0(km, t, self))
+        &*& lifetime_inclusion(a, km) == true;
+    /* Todo: This inclusion must be generated automatically by translator based on reference and its target lifetimes.
+       The target lifetimes always outlive reference lifetime out of compiler guarantees of wellformedness of types.
+    */
+    @*/
+    //@ ens thread_token(t) &*& [qa]lifetime_token(a) &*& full_borrow(a, u32_full_borrow_content(t, result));
+    {
+        //@ let kstrong = open_full_borrow_strong(a, MutexGuardU32_full_borrow_content0(km, t, self), qa/2);
+        //@ open MutexGuardU32_full_borrow_content0(km, t, self)();
+        //@ open MutexGuardU32_own(km)(t, ?lock);
+
+        // We need `(*lock).data |-> _` to get ptr provenance info
```

Comment:
Could you instead just add the `ptr_within_range(...)` fact in the MutexGuardU32_own predicate?

Reply:
Adding `pointer_within_limits`  to `MutexGuardU32_own` means `MutexGuardU32_new` should provide it which in turn it means either this method should open the full borrow or should receive it by contract. the latter means `MutexU32::lock`should open the full borrow to get that fact. So I think it would not save us proof lines for now. However, If we put the fact in `MutexGuardU32_own` it will help when we implement the `deref` method later.

I have type matching problem writing `field_pte(lock, ...)` because it needs a value of `pointer` type in the position of `lock` and `lock` is just `*MutexU32`.

Reply:
One can cast a value `v` of type `T *` to type `pointer` using a cast: `(pointer)v`. But we can look at this later.